Source/Csla.Shared/Server/DataPortal.cs

```diff
@@ -752,6 +740,18 @@ private void SetContext(DataPortalContext context)
 
     private void ClearContext(DataPortalContext context)
     {
+#if !NET40 && !NET45
+      if (_oldLocation == ApplicationContext.LogicalExecutionLocations.Client)
```

Comment:
I don't think this is a valid assumption when the logical server-side code is running on a physical app server.

Reply:
Csla.ApplicationContext.LogicalExecutionLocation returns LogicalExecutionLocations.Client when the value is not present in the ContextDictionary:
```
    public static LogicalExecutionLocations LogicalExecutionLocation
    {
      get
      {
        object location = LocalContext.GetValueOrNull("__logicalExecutionLocation");
        if (location != null)
          return (LogicalExecutionLocations)location;
        else
          return LogicalExecutionLocations.Client;
      }
    }
```
 It happens when the first DataPortal call reach Csla.Server.DataPortal.SetContext and this value is stored in "_oldLocation":

    ```
private void SetContext(DataPortalContext context)
    {
      _oldLocation = Csla.ApplicationContext.LogicalExecutionLocation;
      ApplicationContext.SetLogicalExecutionLocation(ApplicationContext.LogicalExecutionLocations.Server);

      if (!context.IsRemotePortal && ApplicationContext.WebContextManager != null && !ApplicationContext.WebContextManager.IsValid)
        ApplicationContext.SetContext(context.ClientContext, context.GlobalContext);

```

This way we can use _oldLocation to know if the call is about to leave the logical server. Am I correct?

Reply:
Hmm. That is an unintentional behavior. Not that I plan to change it, but the behavior is a coincidence, not part of any specific design.

Now it becomes a pre-requisite for some functionality, and so stops being unintentional - does that make sense?

That's ok, but requires a unit test to confirm that this is the behavior (that the value starts out as "client"), so if it ever does change in the future we know that is bad.

Reply:
Totally makes sense. It is unintentional but excellent behavior: it created an easy way to check if the code is about to leave the server. 
I created tests to check the initial value of ApplicationContext.LogicalExecutionLocation.

Reply:
Thank you @Eduardo-Micromust !